File: Source/NSString+PecuniaAdditions.m
/**
 * Copyright (c) 2012, 2014, Pecunia Project. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; version 2 of the
 * License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

#import "NSString+PecuniaAdditions.h"
#import "Mathematics.h"
#import "MessageLog.h"

#import "ZipFile.h"
#import "FileInZipInfo.h"
#import "ZipReadStream.h"

@implementation NSString (PecuniaAdditions)

static NSData *receivedTmpData = nil;

/**
 * Returns a new string instance containing the given data as a string (assuming UTF8 encoding).
 * If the given data cannot be converted to an UTF8 string it is converted to a hex string.
 * Returns an empty string if data is nil or empty.
 */
+ (NSString *)stringWithData: (NSData *)data {
    if (data == nil) {
        return [NSString string];
    }

    // issue with received data: if an UTF8 string is cut in between on byte level, its parts do not necessarily represent
    // proper UTF8 strings, e.g. if the cut is between the 2 bytes of a single character
    // we therefore have to look if the combination of 2 or several received data blocks build a valid UTF8 string.
    if (receivedTmpData != nil) {
        NSMutableData *mData = [NSMutableData dataWithData: receivedTmpData];
        [mData appendData: data];
        data = mData;
    }

    NSString *result = [[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding];
    if (result == nil) {
        receivedTmpData = data;
        return @"";
    }
    receivedTmpData = nil;
    return result;
}

- (NSAttributedString *)attributedStringWithFont: (NSFont *)font {
    NSDictionary *attrs = @{
        NSFontAttributeName: font
    };
    NSAttributedString *result = [[NSAttributedString alloc] initWithString: self attributes: attrs];
    return result;
}

- (NSString *)stringByRemovingWhitespaces: (NSString *)s {
    if (s == nil) {
        return nil;
    }
    NSString *result = @"";
    NSArray  *components = [s componentsSeparatedByCharactersInSet: [NSCharacterSet whitespaceAndNewlineCharacterSet]];
    for (NSString *str in components) {
        result = [result stringByAppendingString: str];
    }
    return result;
}

/**
 * Parses the content of the receiver (which must be CSV data) into an array of arrays (rows with columns).
 * Correctly handles quoted fields and fields containing line breaks.
 */
- (NSArray *)csvRowsWithSeparator: (NSString *)separator {
    NSMutableArray *rows = [NSMutableArray array];

    NSMutableCharacterSet *newlineCharacterSet = [NSMutableCharacterSet whitespaceAndNewlineCharacterSet];
    [newlineCharacterSet formIntersectionWithCharacterSet: [[NSCharacterSet whitespaceCharacterSet] invertedSet]];

    NSMutableCharacterSet *importantCharactersSet =
        [NSMutableCharacterSet characterSetWithCharactersInString: [NSString stringWithFormat: @"%@\"", separator]];
    [importantCharactersSet formUnionWithCharacterSet: newlineCharacterSet];

    // Construct a characterset for skipping whitespaces between fields, but exclude the separator
    // (which might be a whitespace too).
    NSMutableString *skipChars = [NSMutableString string];
    if ([separator rangeOfString: @" "].length == 0) {
        [skipChars appendString: @" "];
    }
    if ([separator rangeOfString: @"\t"].length == 0) {
        [skipChars appendString: @"\t"];
    }
    NSCharacterSet *skipCharacterSet = [NSCharacterSet characterSetWithCharactersInString: skipChars];

    NSScanner *scanner = [NSScanner scannerWithString: self];
    [scanner setCharactersToBeSkipped: nil];
    while (![scanner isAtEnd]) {
        BOOL insideQuotes = NO;
        BOOL finishedRow = NO;

        NSMutableArray  *columns = [NSMutableArray arrayWithCapacity: 10];
        NSMutableString *currentColumn = [NSMutableString string];
        while (!finishedRow) {
            NSString *tempString;
            if ([scanner scanUpToCharactersFromSet: importantCharactersSet intoString: &tempString]) {
                [currentColumn appendString: tempString];
            }

            if ([scanner isAtEnd]) {
                if (![currentColumn isEqualToString: @""]) {
                    [columns addObject: currentColumn];
                }
                finishedRow = YES;
            } else if ([scanner scanCharactersFromSet: newlineCharacterSet intoString: &tempString]) {
                if (insideQuotes) {
                    // Add line break to column text
                    [currentColumn appendString: tempString];
                } else {
                    // End of row
                    if (![currentColumn isEqualToString: @""]) {
                        [columns addObject: currentColumn];
                    }
                    finishedRow = YES;
                }
            } else if ([scanner scanString: @"\"" intoString: nil]) {
                if (insideQuotes && [scanner scanString: @"\"" intoString: nil]) {
                    // Replace double quotes with a single quote in the column string.
                    [currentColumn appendString: @"\""];
                } else {
                    // Start or end of a quoted string.
                    insideQuotes = !insideQuotes;
                }
            } else if ([scanner scanString: separator intoString: nil]) {
                if (insideQuotes) {
                    [currentColumn appendString: separator];
                } else {
                    // This is a column separating delimiter.
                    [columns addObject: currentColumn];
                    currentColumn = [NSMutableString string];

                    [scanner scanCharactersFromSet: skipCharacterSet intoString: nil];
                }
            }
        }
        if ([columns count] > 0) {
            [rows addObject: columns];
        }
    }

    return rows;
}

/**
 * Interprets the content as XML and inserts line breaks and whitespaces for better readability.
 */
- (NSString *)formatXML {
    NSError       *error;
    NSXMLDocument *document = [[NSXMLDocument alloc] initWithXMLString: self options: NSXMLNodePreserveAll error: &error];
    if (error != nil) {
        return self;
    }
    return [document XMLStringWithOptions: NSXMLNodePrettyPrint];
}

- (BOOL)hasSubstring: (NSString *)substring {
    if (substring != nil) {
        NSRange r = [self rangeOfString: substring];
        return r.location != NSNotFound;
    }
    return NO;
}

- (NSString *)stringByEscapingXmlCharacters {
    NSCharacterSet *cs = [NSCharacterSet characterSetWithCharactersInString: @"&<>'"];
    NSRange        range = [self rangeOfCharacterFromSet: cs];
    if (range.location == NSNotFound) {
        return self;
    }
    NSString *res = [self stringByReplacingOccurrencesOfString: @"&" withString: @"&amp;"];
    res = [res stringByReplacingOccurrencesOfString: @"<" withString: @"&lt;"];
    res = [res stringByReplacingOccurrencesOfString: @">" withString: @"&gt;"];
    res = [res stringByReplacingOccurrencesOfString: @"'" withString: @"&apos;"];
    return res;
}

@end

File: Source/NSString+PecuniaAdditions.swift
/**
* Copyright (c) 2014, 2015, Pecunia Project. All rights reserved.
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License as
* published by the Free Software Foundation; version 2 of the
* License.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
* 02110-1301  USA
*/

import Foundation

extension NSString {
    /**
    * Processes the content of the receiver and creates a new string with a more natural appearance, if possible.
    * This mostly involves truecasing words.
    * This process depends on a large word library, so it's an optional feature. If the user decided
    * to disable it we do a simple capitalization.
    */
    public func stringWithNaturalText() -> NSString {
        if !WordMapping.wordMappingsAvailable {
            // While our word list is being loaded return a simple capitalized string.
            return self.capitalized as NSString;
        }

        let context = MOAssistant.shared().context;
        let request = NSFetchRequest<NSFetchRequestResult>();
        request.entity = NSEntityDescription.entity(forEntityName: "WordMapping", in: context!);

        let result = NSMutableString();
        let range = NSMakeRange(0, self.length);
        self.enumerateLinguisticTags(in: range, scheme: NSLinguisticTagSchemeLexicalClass,
            options: NSLinguisticTagger.Options(rawValue: 0), orthography: nil,  using:
            { (tag: String, tokenRange: NSRange, sentenceRange: NSRange, stop: UnsafeMutablePointer<ObjCBool>) -> () in
                let item : NSString = self.substring(with: tokenRange) as NSString;
                if tag == NSLinguisticTagOtherWhitespace {
                    // Copy over any whitespace.
                    if result.length > 0 {
                        result.append(item as String);
                    }
                } else {
                    // Not a whitespace. See if that is a known word.
                    // TODO: needs localization.
                    var key = item.stringWithNormalizedGermanChars().lowercased;
                    key = key.replacingOccurrences(of: "'", with: "");
                    let predicate = NSPredicate(format: "wordKey = '\(key)'");
                    request.predicate = predicate;
                    do {
                        let mappings = try context?.fetch(request);
                        if ((mappings?.count)! > 0) {
                            let mapping = mappings?[0] as! WordMapping;
                            let word: NSString = mapping.translated as NSString;

                            // If the original word contains lower case characters then it was probably
                            // already in true case. We only use the lookup then for replacing diacritics/sharp-s.
                            if item.rangeOfCharacter(from: CharacterSet.lowercaseLetters).length > 0 {
                                result.append(item.substring(to: 1));
                                result.append(word.substring(from: 1));
                            } else {
                                // Make the first letter upper case if it is the first entry.
                                // Don't touch the other letters, though!
                                if result.length == 0 {
                                    let firstLetter: NSString = word.substring(to: 1) as NSString;
                                    result.append(firstLetter.capitalized);
                                    result.append(word.substring(from: 1));
                                } else {
                                    result.append(word as String);
                                }
                            }
                        } else {
                            result.append(item as String);
                        }
                    }
                    catch {
                        // Do nothing in case of an error. Should never happen, but just in case.
                    }
                }
            }
        );
        return result;
    }

    /**
    * Converts umlauts and ß in the receiver to a decomposed form.
    */
    public func stringWithNormalizedGermanChars() -> NSString {
        var result = self.decomposedStringWithCanonicalMapping;
        result = result.replacingOccurrences(of: "ß", with: "ss");
        return result.replacingOccurrences(of: "\u{0308}", with: "e") as NSString; // Combining diaresis.
    }
}
